scripts/create_singularities

```diff
@@ -40,264 +32,279 @@
-
-# Docker image for simple_workflow
-generate_singularity_for_docker_image "ReproNim/simple_workflow" "repronim"
-# One time run! specify nonesense to not match anything else
-# add_singularity_versioned_image "ReproNim/containers" "repronim" "repronim-ptb-3--3.0.15.20190401.dfsg1-1+nd100"
-# TODO: it resulted in repronim-containers--repronim-ptb-3--3.0.15.20190401.dfsg1-1+nd100.sing 
-#       image filename since used that prefix... renaming manually for now
-
-# neuronets
-generate_singularity_for_docker_image "neuronets/kwyk" "neuronets"
+from __future__ import annotations
+from dataclasses import dataclass
+import logging
+import os
+from pathlib import Path
+import re
+import subprocess
+import sys
+from typing import Any, Optional
+import requests
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class Builder:
+    images_dir: Path
+    tmp_dir: Path
+    githubids: set[str]
+
+    def runcmd(self, *args: str, **kwargs: Any) -> subprocess.CompletedProcess:
+        kwargs.setdefault("check", True)
+        kwargs["cwd"] = str(self.images_dir)
+        kwargs["env"] = {
+            **os.environ,
+            "TMPDIR": str(self.tmp_dir),
+            "SINGULARITY_TMPDIR": str(self.tmp_dir),
+        }
+        return subprocess.run(args, **kwargs)
+
+    @staticmethod
+    def get_last_docker_version_tag(dh: str) -> Optional[str]:
+        r = requests.get(f"https://registry.hub.docker.com/v2/repositories/{dh}/tags")
+        r.raise_for_status()
+        versions = [res["name"] for res in r.json()["results"]]
+        if len(versions) > 1:
+            # select only the ones which seems to be semantic and/or
+            # master/latest.  Some release alpha releases, so probably would
+            # skip those for now
+            good_versions = [
+                re.sub(r"^(([Vv]|version-|)([0-9]{1,10}\..*))", r"\3@\1", v)
+                for v in versions
+                if not (
+                    re.search(r"[ab][0-9]+$", v)
+                    or re.search(r"rc[0-9]*$", v)
+                    or "master" in v
+                )
+            ]
+            return max(good_versions, key=version_key, default=None)
+        elif not versions:
+            log.info(" %s no version. Tags: %s", dh, " ".join(versions))
+            return None
+        else:
+            return versions[0]
+
+    def create_singularity_file(
+        self, dh: str, version_tag: str, version_pure: str, family: str
+    ) -> SingularityFile:
+        imagename = get_imagename(dh, family, version_pure)
+        outdir = self.images_dir / family
+        outdir.mkdir(parents=True, exist_ok=True)
+        singfile = outdir / f"Singularity.{imagename}"
+        # log.info("%s \t from \t %s:%s", singfile, dh, version_tag)
+        # Do not recreate the file ATM. Since we changed the env vars we
+        # define, we don't want to recreate it if already present
+        if not singfile.exists():
+            singfile.write_text(
```

Comment:
Let's just use one mutliline `f"""\` string here to avoid all per line decorations

Reply:
Done.